### PR TITLE
Shared + map

### PR DESCRIPTION
**Pull request by arnauddorgans**
Adding a `map` method to both Shared and SharedReader

**Motivation**

Avoid sharing parent feature logic with a child.

**E.g.**

```swift
// Feature
@Shared var selectedID: ID 
var children: IdentifiedArrayOf<Child>

// Child
@SharedReader var isSelected: Bool 

// Child Initial State
.init(isSelected: $selectedID.map { $0 == id })
```

**Note**

It could also be cool to add a writable map method, but I don't see the use case today

### CODE REVIEW

**Comment by mbrandonw** [comment]
Hi @arnauddorgans, thanks for exploring this! 

However, it should not be necessary. `Shared` and `SharedReader` already support a mapping-like operation via dynamic member lookup, much how `Binding` works. So, you can just move your logic to a property or subscript and it should just work:

```swift
extension Child.ID {
  subscript(equals other: Self) -> Bool { self == other }
}

$selectedID[equals: id]  // SharedReader<Bool>
```

Can you give that a shot and let us know how it works out?

**Comment by arnauddorgans** [comment]
@mbrandonw hey, yes it works today I have this extension in my app 

```swift
public extension Equatable {
    subscript(isEqualTo other: Self) -> Bool {
        self == other
    }
}
```

But it feels so unnatural to write a subscript to be able to use it with the library
I think it would be more comfortable to handle `map` methods, like Sequence, AsyncSequence or Publisher

Plus, we can't extend `Any` or `Sendable` so we can't share custom `Sharing` maps across types

**Comment by stephencelis** [comment]
@arnauddorgans One wrinkle is that the `@Shared` property wrapper's underlying reference's identity depends on the source shared key, which is a unique, hashable value, and then any transformations/projections along the way are also hashable, since they rely on hashable key paths.

This means that these 2 shared references are equal:

```swift
@Shared(.key) var value1
@Shared(.key) var value2

$value1.prop == $value2.prop  // true
```

But because the transform function above is _not_ hashable, the implementation uses the object identity, instead, which will not be shared:

https://github.com/pointfreeco/swift-sharing/pull/138/files#diff-c5b71fccf18418e74ad127f0ae15e15ae16773156494acece9fa6d296edb0da0R672-R673

This means that mapped references will never be equal:

```swift
@Shared(.key) var value1
@Shared(.key) var value2

$value1.map { $0.prop } == $value2.map { $0.prop }  // true
```

It's possible this isn't a big deal with readers, but at the same time it's a subtle enough difference that should be carefully considered. I wonder if you can come up with some more edge cases to test and explore considering this mismatch?

One other thing that would be good is to get coverage on some actual observation, since that's probably the entire point of such a shared reader. Both tests on `observe` firing when we expect it to, as well as tests on `publisher` emitting with the transformed value would be good to have.

**Comment by arnauddorgans** [comment]
@stephencelis I don't think the object comparison is a big deal, it is already done for `_BoxReference` and `_PersistentReference`

Plus, the Hashable constraint is annoying because it only constrains us to use `Hashable` parameters.

Screenshot 2025-03-26 at 03 56 30


I'll try to write some additional tests when I have the time 👍

**Comment by stephencelis** [comment]
@arnauddorgans The difference is that `_Box` and `_PersistentReference` are the _root_ references and thus there is only a single one shared amongst all projected children.

In the case of `map`, every time a `map` happens a new object is created, even though they may reference the same underlying data. Again, this may be OK, but I haven't yet fully wrapped my head around the difference yet to see if there's a problem :smile:

**Comment by arnauddorgans** [comment]
@stephencelis in my last commit, I deleted the `Equatable` conformance when the reference is not writable. 
It wasn't needed, so the reference check that I wrote has been deleted as map is only available for `SharedReader`

This way I think that everything is working as all the existing shared references

So it is totally legit that
```swift
@Shared(value: 0) var base: Int
@SharedReader var lhs: Int
@SharedReader var rhs: Int
_lhs = $base.map { $0 * 2 }
_rhs = $base.map { $0 * 3 }
#expect(lhs == rhs)
#expect($lhs == $rhs)
    
$base.withLock { $0 += 1 }
#expect(lhs != rhs)
#expect($lhs != $rhs)
```

**Comment by stephencelis** [comment]
While we think on this, @jshier I believe you were asking for something like this?

**Comment by jshier** [comment]
Yes, I needed a simple way to create a `SharedReader` from a parent `Shared` value using a query to parent opaque blob, whether on the fly or through a protocol abstraction. Only way I got it to work was to create a computed property in an extension on the parent type, which made it too unergonomic to use for dozens of different results across the codebase. I abandoned the approach for that reason and we just use the parent blob, which is awkward for tests but otherwise works fine. 

That particular use case could also use a better solution for exposing custom APIs out `Shared` values, or their wrappers.

File: Sources/Sharing/Internal/Reference.swift [hunk 1]
```diff
@@ -529,15 +529,15 @@ extension _ManagedReference: MutableReference, Equatable where Key: SharedKey {
   }
 }
 
-final class _AppendKeyPathReference<
```
Comment (stephencelis) [inline Sources/Sharing/Internal/Reference.swift:532]: Let's leave the current hierarchy as is. While tempting to abstract here, for now I think it's a little easier to maintain as a brand new `_ReadClosureReference` conformance.

File: Sources/Sharing/Shared.swift [hunk 2]
```diff
@@ -223,6 +223,15 @@ public struct Shared<Value> {
       reference = newValue.reference
     }
   }
+  
+  /// Returns a read-only shared reference to the resulting value of a given closure.
+  ///
+  /// - Returns: A new read-only shared reference.
+  public func map<Member>(
```
Comment (stephencelis) [inline Sources/Sharing/Shared.swift:230]: Let's call this `read` instead. While it is very `map`-like I think it might be confusing with the type changing `Shared.map -> SharedReader`. A domain-specific word is maybe easier to understand.

**Review by stephencelis — COMMENTED** [review]
@arnauddorgans Chatted through this again today and I think we're down to include this functionality. Just a couple changes and I think we'll be good to go! Let us know what you think :smile:

File: Sources/Sharing/Shared.swift (hunk 2 above)
Comment (arnauddorgans) [reply Sources/Sharing/Shared.swift:230]: you think so? we already have an existing behavior dynamicMemberLookup with non writable keyPath
```swift
  public subscript<Member>(
    dynamicMember keyPath: KeyPath<Value, Member>
  ) -> SharedReader<Member> {
    SharedReader(self)[dynamicMember: keyPath]
  }
```

**Review by stephencelis — COMMENTED** [review]
@arnauddorgans Updated reference looks good! Want to make a final change of renaming `map` to `reader` and we can get this merged?

File: Sources/Sharing/Shared.swift (hunk 2 above)
Comment (arnauddorgans) [reply Sources/Sharing/Shared.swift]: done ✅

File: Sources/Sharing/Shared.swift [hunk 3]
```diff
@@ -223,6 +223,15 @@ public struct Shared<Value> {
...
   }
+  
+  /// Returns a read-only shared reference to the resulting value of a given closure.
+  ///
+  /// - Returns: A new read-only shared reference.
+  public func reader<Member>(
+    _ body: @escaping @Sendable(Value) -> Member
+  ) -> SharedReader<Member> {
+    SharedReader(self).map(body)
+  }
```
Comment (stephencelis) [inline Sources/Sharing/Shared.swift:234]:
```suggestion
  public func read<Member>(
    _ body: @escaping @Sendable(Value) -> Member
  ) -> SharedReader<Member> {
    SharedReader(self).read(body)
  }
```

File: Sources/Sharing/SharedReader.swift [hunk 4]
```diff
@@ -155,6 +155,20 @@ public struct SharedReader<Value> {
       reference = newValue.reference
     }
   }
+  
+  /// Returns a read-only shared reference to the resulting value of a given closure.
+  ///
+  /// - Returns: A new shared reader.
+  public func map<Member>(
```
Comment (stephencelis) [inline Sources/Sharing/SharedReader.swift:162]:
```suggestion
  public func read<Member>(
```

File: Tests/SharingTests/EquatableTests.swift [hunk 5]
```diff
@@ -32,4 +32,18 @@ struct EquatableTests {
     #expect(lhs == rhs)
     #expect($lhs == $rhs)
   }
+  
+  @Test func mapReader() {
+    @Shared(value: 0) var base: Int
+    @SharedReader var lhs: Int
+    @SharedReader var rhs: Int
+    _lhs = $base.reader { $0 * 2 }
+    _rhs = $base.reader { $0 * 3 }
```
Comment (stephencelis) [inline Tests/SharingTests/EquatableTests.swift:41]:
```suggestion
    _lhs = $base.read { $0 * 2 }
    _rhs = $base.read { $0 * 3 }
```

File: Tests/SharingTests/SharedTests.swift [hunk 6]
```diff
@@ -87,6 +87,22 @@ import Testing
 
       #expect(id == 42)
     }
+    
+    @Test func mapReader() {
+      @Shared(value: 0) var count
+      @SharedReader var isZero: Bool
+      _isZero = $count.reader { $0 == 0 }
```
Comment (stephencelis) [inline Tests/SharingTests/SharedTests.swift:94]:
```suggestion
      _isZero = $count.read { $0 == 0 }
```

**Review by stephencelis — COMMENTED** [review]
One last bike shed!

**Review by stephencelis — APPROVED** [review]
